python/pfs/utils/optimize_black_dots.py

```diff
@@ -0,0 +1,482 @@
+                    else:
+                        predicted_position_x = np.empty(len(x_measurments))
+                        predicted_position_x[:] = np.NaN
+                        predicted_position_y = np.empty(len(y_measurments))
+                        predicted_position_y[:] = np.NaN
+                else:
+                    if residuals_y[0] < 0.5 and residuals_x[0] < 0.5\
+                            and ~((np.sum(np.diff(idx)) == 1) and (r_distance_predicted > 1.5)):
+                        predicted_position_x = poly_deg_x(np.arange(0, len(x_measurments)))
+                        predicted_position_y = poly_deg_y(np.arange(0, len(y_measurments)))
+                    else:
+                        predicted_position_x = np.empty(len(x_measurments))
+                        predicted_position_x[:] = np.NaN
+                        predicted_position_y = np.empty(len(y_measurments))
+                        predicted_position_y[:] = np.NaN
+
+                # create lists which contain observed and predicted data
+                actual_position_x_observed = mcs_data_all[0, i][idx]
+                actual_position_y_observed = mcs_data_all[1, i][idx]
+
+                predicted_position_x_not_observed = predicted_position_x[~idx]
+                predicted_position_y_not_observed = predicted_position_y[~idx]
+
+                list_of_actual_position_x_observed.append(actual_position_x_observed)
+                list_of_actual_position_y_observed.append(actual_position_y_observed)
+
+                list_of_predicted_position_x_not_observed.append(predicted_position_x_not_observed)
+                list_of_predicted_position_y_not_observed.append(predicted_position_y_not_observed)
+            except (TypeError, ValueError):
+                # if the process failed at some points, do not consider this fiber in the later stages
+                list_of_actual_position_x_observed.append(np.nan)
+                list_of_actual_position_y_observed.append(np.nan)
+
+                list_of_predicted_position_x_not_observed.append(np.nan)
+                list_of_predicted_position_y_not_observed.append(np.nan)
+
+        prepared_observations_and_predictions = [list_of_actual_position_x_observed,
+                                                 list_of_actual_position_y_observed,
+                                                 list_of_predicted_position_x_not_observed,
+                                                 list_of_predicted_position_y_not_observed]
+
+        return prepared_observations_and_predictions
+
+    def rotate(self, point, origin, degrees):
+        """rotate the point around the origin
+
+        gets called by `new_position_of_dot`
+
+        Parameters
+        ----------
+        point: `tuple`
+            x, y coordinate of the point to rotate
+        origin: `tuple`
+            x, y coordinate of the point around which to rotate
+        degrees: `float`
+            number of degrees to rotate
```

Comment:
I suggest changing the variable name to 'angle' or similar, and specify in the docString that the units are degrees.